src/url-loader/indirect-url-resolver.ts

```diff
@@ -0,0 +1,122 @@
+/**
+ * @license
+ * Copyright (c) 2017 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {FileRelativeUrlBrand, PackageRelativeUrl} from '../index';
+import {FileRelativeUrl, ResolvedUrl} from '../model/url';
+
+import {FsUrlResolver} from './fs-url-resolver';
+
+type RuntimeUrl = string&RuntimeUrlBrand;
+
+
+// Declare these as classes rather than interfaces so that the properties
+// can be private.
+declare class RuntimeUrlBrand extends FileRelativeUrlBrand {
+  private RuntimeUrlBrand: never;
+}
+
+/**
+ * A URL resolver for very large codebases where source files map in an
+ * arbitrary but predetermined fashion onto URL space.
+ *
+ * It also separates the root directory – the root of all source code that's
+ * legal to load – from the package directory, which is how the user refers to
+ * files on the CLI or the IDE.
+ */
+export class IndirectUrlResolver extends FsUrlResolver {
+  private readonly runtimeUrlToResolvedUrl:
+      ReadonlyMap<RuntimeUrl, ResolvedUrl>;
+  private readonly resolvedUrlToRuntimeUrl:
+      ReadonlyMap<ResolvedUrl, RuntimeUrl>;
+
+  /**
+   * @param rootPath All loadable source code must be a descendent of this
+   *     directory. Should be the same as FsUrlLoader.
+   * @param packagePath The base directory for paths from the user. Usually the
+   *     current working directory.
+   * @param urlToFilesystemMap Maps the runtime URL space to the paths for those
+   *     files on the filesystem. The runtime URLs should all be relative paths
+   *     from the same base url. The filesystem paths should all be relative
+   *     paths from `rootPath`.
+   */
+  constructor(
+      rootPath: string, packagePath: string,
+      indirectionMap: Map<string, string>) {
+    super(packagePath);
+
+    const rootResolver = new FsUrlResolver(rootPath);
+    const urlspaceToFilesystem = new Map<RuntimeUrl, ResolvedUrl>();
```

Comment:
not sure you really need the locals

Reply:
The fields are ReadonlyMaps, documenting that they're not changed after creation.